Fix bug in `pallet-evm` genesis account creation

tgmichel opened the pull request:
Inserting to `AccountCodes` directly without increasing Account reference (`inc_consumers`) leads to errors if that account is removed later on using the `remove_account` function:

```
pub fn remove_account(address: &H160) {
	if <AccountCodes<T>>::contains_key(address) {
		let account_id = T::AddressMapping::into_account_id(*address);
		let _ = frame_system::Pallet::<T>::dec_consumers(&account_id);
	}

	<AccountCodes<T>>::remove(address);
	<AccountStorages<T>>::remove_prefix(address, None);
}
```

Above `AccountCodes` key will exists, but never increased the consumers:

```
/// Decrement the reference counter on an account. This *MUST* only be done once for every time
/// you called `inc_consumers` on `who`.
pub fn dec_consumers(who: &T::AccountId)
```

In addition to that, in order to increase the consumer reference count the Account must have increased the providers. Otherwise it fails with `DispatchError::NoProviders`.

JoshOrndorff reviewed (approved):
Is it possible to write a test for this?

tgmichel commented:
> Is it possible to write a test for this?

Thanks Joshy, I added the tests and also they surfaced another problem: in order to increase the consumer reference, the provider reference must be greater than zero. So I added increasing / decreasing provider reference in the create and remove account functions.

sorpaas reviewed (changes requested):
`AccountCodes` is only a consumer of the refcount, but not a provider. We determine existence of EVM account not by `AccountCodes`, but by the Substrate account information. This currently has to be the case because EVM's `nonce` corresponds to Substrate's `nonce`. In addition, the current change is unsafe without a migration.

You can call `inc_providers` in genesis builder. This will create a dangling pointer. Given that it's genesis it shouldn't be of problem.

tgmichel commented:
> `AccountCodes` is only a consumer of the refcount, but not a provider. We determine existence of EVM account not by `AccountCodes`, but by the Substrate account information. This currently has to be the case because EVM's `nonce` corresponds to Substrate's `nonce`. In addition, the current change is unsafe without a migration.
> 
> You can call `inc_providers` in genesis builder. This will create a dangling pointer. Given that it's genesis it shouldn't be of problem.

Changed, so we assume all accounts have at least 1 provider reference except the genesis accounts?

sorpaas commented:
@tgmichel Can you change this PR to use `sufficients`? We can work on the migration later.
